File: packages/gbnf-experiment/data/20260909T130706Z_31ac2789/ported_implementation/src/grammar_graph/get_parent_stack_id.ts
import { Color } from './colorize.ts';

import type { Colorize } from './grammar_graph_types.ts';
import type { GraphPointer } from './graph_pointer.ts';

export const getParentStackId = (pointer: GraphPointer, col: Colorize): string => {
  const stackIds: string[] = [];
  let parent: GraphPointer | undefined | null = pointer.parent;
  while (parent) {
    stackIds.push(
      `${parent.node.meta.stackId},${parent.node.meta.pathId},${parent.node.meta.stepId}`,
    );
    parent = parent.parent;
  }
  const arrow = col('<-', Color.GRAY);
  return stackIds.map((stackId) => col(stackId, Color.RED)).join(arrow);
};

export const get_parent_stack_id = getParentStackId;

File: proposed-projects/round-trip-experiment/reverse/20260909T020947Z_ffb57550/20260910T054605Z_51904c27/ported_implementation/gbnf/grammar_graph/get_parent_stack_id.py
from __future__ import annotations

from typing import TYPE_CHECKING, Callable, List

from .colorize import Color

if TYPE_CHECKING:  # pragma: no cover
    from .graph_pointer import GraphPointer


def get_parent_stack_id(pointer: "GraphPointer", col: Callable) -> str:
    stack_ids: List[str] = []
    parent = pointer.parent
    while parent:
        meta = parent.node.meta
        stack_ids.append(f"{meta.stack_id},{meta.path_id},{meta.step_id}")
        parent = parent.parent
    arrow = col("<-", Color.GRAY)
    return arrow.join(col(stack_id, Color.RED) for stack_id in stack_ids)
